### ISSUE TITLE
Stricter test modules

### ISSUE THREAD

**Issue by JonathanWoollett-Light** [rust-lang/rust-clippy#10506]
### What it does

It is easier to illustrate this as a lint formed of multiple lints.

- [x] `test_outside_tests_module`: Warns when a unit test is outside a tests module.

- [ ] ~~`unflagged_tests_module`: Warns when a module named `tests` is not flagged with `#[cfg(test)]`.~~

- [ ] `non_eof_tests_module`: Warns when a module named `tests` is not at the end of a file.

### Lint Name

strict_tests_module

### Category

style, pedantic

### Advantage

- Enforces the idiomatic style of unit tests.
- Allows tools like [`grcov`](https://github.com/mozilla/grcov) to correctly exclude unit tests from coverage with `--excl-start "mod tests"`.

### Drawbacks

- Tools like [`bindgen`](https://github.com/rust-lang/rust-bindgen) often include tests outside test modules. This would warn on this case.

### Example

```rust
#[test]
fn my_test() { /* ... */ }
#[cfg(test)]
mod tests { /* ... */ }
```

Could be written as:

```rust
#[cfg(test)]
mod tests {
    #[test]
    fn my_test() { /* ... */ }
    /* ... */
}
```

---

```rust
mod tests { /* ... */ }
```

Could be written as:

```rust
#[cfg(test)]
mod tests { /* ... */ }
```

---

```rust
#[cfg(test)]
mod tests { /* ... */ }
fn an_item_after() { /* ... */ }
```

Could be written as:

```rust
fn an_item_after() { /* ... */ }
#[cfg(test)]
mod tests { /* ... */ }
```

**Comment by blyxyas** [comment]
I think this issue would be better formatted as a [task list](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/about-task-lists) instead of a numbered list, just for clarity.

@rustbot claim

**Comment by JonathanWoollett-Light** [comment]
@blyxyas Changed to task list.

**Comment by Manishearth** [comment]
See https://github.com/rust-lang/rust-clippy/pull/10515#issuecomment-1478400248: `unflagged_test_modules` is probably not worth it.

The other two sound useful though.

**Comment by Manishearth** [comment]
done

**Comment by smoelius** [comment]
I interpreted the lint mean: a test and a test module within the same enclosing file/module (which I think would be unusual).

In other words, I interpreted the `mod tests` in the example to be necessary.

**Comment by Alexendoo** [comment]
Definitely not as popular but [still 1.9k results](https://github.com/search?q=lang%3Ars+%2F%5E%23%5C%5Btest%5B%5Cs%5CS%5D*%5Cnmod+tests+%2F&type=code)

Looking a few of them it seems common to put a top level test fn right next to the thing it's testing even if there's a test module later in the file, e.g. https://github.com/tokio-rs/axum/blob/b314cd9bc23f6eeb62a1f11a80b95a1720c7816d/axum/src/extract/request_parts.rs#L174

**Comment by blyxyas** [comment]
`non_eof_test_modules` isn't really a clear name (It assumes that the user knows what EOF means), maybe `items_after_test_module`?

**Comment by Alexendoo** [comment]
For things after the test module I [found](https://github.com/search?q=lang%3Ars+%2F%5Emod+tests%3F+%5B%5Cs%5CS%5D*%5Cn%7D%5B%5Cs%5CS%5D*%5Cn%5CS%2F&type=code&p=1) a couple cases that seem like they'd be worth ignoring

Another test or benchmark module following the initial test module, e.g. https://github.com/xi-editor/xi-editor/blob/d027769cec36c3b479ead6a5e963fb5a664149a9/rust/rope/src/delta.rs#L887

And test-ish things like static asserts, e.g. https://github.com/rome/tools/blob/763fd98ad9072a17ca2ef7ed29c6a164eedc8a6b/crates/rome_formatter/src/format_element.rs#L379

The second being less clear cut, I could see someone wanting them to go above the tests module

**Comment by blyxyas** [comment]
So the lint should ignore cases where items behind a test module are also test modules.

**Comment by Alexendoo** [comment]
Yeah, though it might not be obvious if they're test modules depending on the cfg setup, a safe bet would be to ignore any modules

**Comment by blyxyas** [comment]
I'm currently waiting on #10543 to be approved before opening a PR about `items_after_test_module` (changes to that lint may apply also to this one), but as the reviewer seems busy, it may take a few days.

**Comment by blyxyas** [comment]
Now that https://github.com/rust-lang/rust-clippy/commit/5d149c5dac439e332f7c73dc3601f08400abe88e is merged, please cross out `tests_outside_test_module`

**Comment by blyxyas** [comment]
@JonathanWoollett-Light Now that #10578 is merged, could you mark *3* and close the issue?
This was a pretty cool issue to resolve

**Comment by blyxyas** [comment]
Could someone close the issue?
